### PR TITLE
This patch adds ppc64le support back into the Jenkins CI Docker controller build

### PR DESCRIPTION
**Pull request by ksalerno99**
The following are the steps I have automated to build Jenkins CI controller and agents for a foreign architecture. My build workstation architecture is amd64.

I use ```docker.io/multiarch/qemu-user-static``` to emulate the build processes only, I do not use full ```qemu-system-ppc64le``` emulation of a complete Operating System. It is unecessary and slow to attempt full emuluation of kernel etc. to build software, a container is much more efficient and you are always using the latest code, versus maintaining a full Operating System.

If the Jenkins infrastructure team is not already doing so, I suggest switching to this containerized method of building foreign target architecures, as these steps work well for s390x and aarch64 (just substitute ```ARCH=ppc64le``` below with ```ARCH=s390x``` or ```ARCH=arm64v8```).

Build times on a modest 2-core 4-thread Intel i5 M520  @2.40GHz system:

The build time for two controller images (debian_jdk11 and rhel_ubi9_jdk17 simultaneously) was 11 minutes and 23.9 seconds.

The build time for the docker-agent was 4 minutes and 30.8 seconds.

The build time for the docker-ssh-agent was 3 minutes and 57.1 seconds.

Step 1: set ARCH environment variable to desired target architecture
```
ARCH=ppc64le
```

Step 2: run privileged qemu-user-static container to set host's kernel binfmt_misc magic to execute foreign arch binaries with appropriate /usr/bin/qemu-ARCH-static interpreter. IMPORTANT: must use ```--credential yes``` parameter.
```
sudo docker run --rm --privileged \
	docker.io/multiarch/qemu-user-static --reset -p yes --credential yes
```

Step 3: create Docker buildx container with target ARCH. Important: you must mount your host's Docker socket ```/var/run/docker.sock``` in the container, so that we can eliminate the need to run the Docker service inside the emulated container.
```
sudo docker run -dt -h docker-buildx-$ARCH --name docker-buildx-$ARCH \
	-v/var/run/docker.sock:/var/run/docker.sock $ARCH/ubuntu
```

Step 4: build Jenkins CI inside target ARCH Docker container with buildx
```
sudo docker exec -ti docker-buildx-$ARCH /bin/bash -c "\
	apt-get update; \
	apt-get install -y --no-install-recommends \
		ca-certificates curl gnupg lsb-release jq make git vim patch; \
	mkdir -m 0755 -p /etc/apt/keyrings; \
	curl -fsSL \
		https://download.docker.com/linux/ubuntu/gpg | \
		gpg --dearmor -o /etc/apt/keyrings/docker.gpg; \
	echo \"deb [arch=\$(dpkg --print-architecture) \
		signed-by=/etc/apt/keyrings/docker.gpg] \
		https://download.docker.com/linux/ubuntu \
		\$(lsb_release -cs) stable\" | \
		tee /etc/apt/sources.list.d/docker.list >/dev/null; \
	apt-get update; \
	apt-get install -y --no-install-recommends \
		docker-ce-cli docker-buildx-plugin docker-compose-plugin"
sudo docker exec -ti docker-buildx-$ARCH /bin/bash -c "\
	cd; \
	git clone https://github.com/jenkinsci/docker.git; \
	git clone https://github.com/jenkinsci/docker-agent.git; \
	git clone https://github.com/jenkinsci/docker-ssh-agent.git; \
	docker buildx version; docker version; \
	cd docker && time make build && \
	cd ../docker-agent && time make build && \
	cd ../docker-ssh-agent && time make build"
```

Step 5 (optional): run Jenkins CI Docker controller and agent containers on Docker host to test (since we shared the ```/var/run/docker.sock``` of the host, all images are accessible from the host that were exported from the buildx container)
```
JAVA_OPTS="-Djava.util.logging.config.file=/var/jenkins_home/log.properties"
JAVA_PROP=data/log.properties
docker images | grep jenkins/jenkins
printf "Choose Tag to run: "
read TAG
mkdir data
sudo chown 1000:1000 data
echo "handlers=java.util.logging.ConsoleHandler" >$JAVA_PROP
echo "jenkins.level=FINEST" >>$JAVA_PROP
echo "java.util.logging.ConsoleHandler.level=FINEST" >>$JAVA_PROP
sudo docker run --rm --privileged \
	docker.io/multiarch/qemu-user-static --reset -p yes --credential yes
sudo docker run -d --name kensqa \
	-p 8080:8080 -p 50000:50000 --restart=on-failure \
	--env JAVA_OPTS="$JAVA_OPTS" \
	-v $(pwd)/data:/var/jenkins_home jenkins/jenkins:$TAG
mkdir docker-agent-workdir
sudo chown 1000:1000 docker-agent-workdir
sudo docker run -dt --name docker-agent --init \
	-v $(pwd)/docker-agent-workdir:/home/jenkins/agent \
	jenkins/agent java -jar /usr/share/jenkins/agent.jar \
	-workDir /home/jenkins/agent
echo | ssh-keygen -t rsa -N ''
mkdir docker-ssh-agent-workdir
sudo chown 1000:1000 docker-ssh-agent-workdir
sudo docker run -d --name docker-ssh-agent \
	-v $(pwd)/docker-ssh-agent-workdir:/home/jenkins/agent -p 9022:22 \
	-e "JENKINS_AGENT_SSH_PUBKEY=$(cat .ssh/id_rsa.pub)" \
	jenkins/ssh-agent
cat .ssh/id_rsa .ssh/id_rsa.pub
sudo docker ps
```

Results:
```
sudo docker images
REPOSITORY                   TAG                     IMAGE ID       CREATED        SIZE
jenkins/ssh-agent            bullseye-jdk11          a9554b89c3b8   11 hours ago   384MB
jenkins/ssh-agent            jdk11                   a9554b89c3b8   11 hours ago   384MB
jenkins/ssh-agent            latest                  a9554b89c3b8   11 hours ago   384MB
jenkins/ssh-agent            latest-bullseye-jdk11   a9554b89c3b8   11 hours ago   384MB
jenkins/ssh-agent            latest-jdk11            a9554b89c3b8   11 hours ago   384MB
jenkins/agent                bullseye-jdk11          8728cac70f74   11 hours ago   369MB
jenkins/agent                jdk11                   8728cac70f74   11 hours ago   369MB
jenkins/agent                latest                  8728cac70f74   11 hours ago   369MB
jenkins/agent                latest-bullseye-jdk11   8728cac70f74   11 hours ago   369MB
jenkins/agent                latest-jdk11            8728cac70f74   11 hours ago   369MB
jenkins/jenkins              2.356                   a04b8f9b2f1a   11 hours ago   510MB
jenkins/jenkins              2.356-jdk11             a04b8f9b2f1a   11 hours ago   510MB
jenkins/jenkins              2.356-rhel-ubi9-jdk17   1272648dbb50   11 hours ago   542MB
ppc64le/ubuntu               latest                  4220c61b3ab7   12 days ago    118MB
multiarch/qemu-user-static   latest                  3539aaa87393   2 months ago   305MB

sudo docker inspect jenkins/jenkins:2.356 | grep Architecture
        "Architecture": "ppc64le",

sudo docker inspect jenkins/jenkins:2.356-rhel-ubi9-jdk17 | grep Architecture
        "Architecture": "ppc64le",

sudo docker inspect jenkins/agent:jdk11 | grep Architecture
        "Architecture": "ppc64le",

sudo docker inspect image jenkins/ssh-agent:jdk11 | grep Architecture
        "Architecture": "ppc64le",
```

### CODE REVIEW

File: docker-bake.hcl [hunk 1]
```diff
@@ -31,6 +31,13 @@ group "linux-s390x" {
   ]
 }
 
+group "linux-ppc64le" {
+  targets = [
+    "debian_jdk11",
+    "rhel_ubi9_jdk17",
```
Comment (timja) [inline docker-bake.hcl:37]: is there a reason you need it on rhel? I would suggest just debian 11 and 17
```suggestion
    "debian_jdk11",
    "debian_jdk17",
```

Comment (ksalerno99) [reply docker-bake.hcl:37]: Hello, Tim. Yes, we have at least one customer actively using the UBI controller image. We can include all three images in the build:

debian_jdk11
debian_jdk17
rhel_ubi9_jdk17

**Review by dduportal — CHANGES_REQUESTED** [review]
This PR looks really good to me and is clearly needed, thanks!

I'm retaining this PR by adding a "Request for change" as merging it as it would rebuild the 2 last LTS and weekly images (as the script would consider them incomplete).

I'll make the ppc64le build for these releases first, with a manual deploy, then will try the build script with the code of this pr in dr-run to confirm it does not want to rebuild releases.

If it goes ok, I'll approve and merge.

**Comment by dduportal** [comment]
Starting preparatory work for building this new image locally. I'll comment the PR following the actions until merge and initial delivery

**Comment by dduportal** [comment]
First step was a failure:

- Trying with the `2.403` as it's not an LTS and not the latest weekly (and should not be rebuilt after Tuesday 16 May) so accidentally  overriding it as a really low impact
- Tried to build and push "only" the `ppc64le` architecture with the following command:

```
JENKINS_SHA=30df8e6c5f04e89e31d53a48cb9df340c22fd3c54c0f14efc494da95c083ffc7
JENKINS_VERSION=2.403
COMMIT_SHA=c5f88b5251a03551829379185390a4439e05e75f
LATEST_WEEKLY= false
LATEST_LTS= false
export JENKINS_VERSION JENKINS_SHA LATEST_WEEKLY LATEST_LTS COMMIT_SHA

docker buildx bake -f docker-bake.hcl --set '*.platform=linux/ppc64le' --push linux-ppc64le
```

=> result is that it rewrites the top-level manifest resulting in the DockerHub only showing the `ppc64le` architecture:

Capture d’écran 2023-05-14 à 12 29 08


=> triggered a new build (in trusted.ci.jenkins.io, the private controller) which rebuilt the `2.403` tag (withotu touching the other builds).

I'm now trying with https://github.com/docker/buildx/blob/master/docs/reference/buildx_imagetools_create.md

**Comment by dduportal** [comment]
Second experiment was successul!

- Started by building and pushing the PPC64 images to my own DockerHub account to obtain signed image with digests:

```
docker login -u dduportal 
# ...
JENKINS_SHA=30df8e6c5f04e89e31d53a48cb9df340c22fd3c54c0f14efc494da95c083ffc7
JENKINS_VERSION=2.403
COMMIT_SHA=c5f88b5251a03551829379185390a4439e05e75f
LATEST_WEEKLY= false
LATEST_LTS= false
export JENKINS_VERSION JENKINS_SHA LATEST_WEEKLY LATEST_LTS COMMIT_SHA
export JENKINS_REPO=dduportal/jenkins

docker buildx bake -f docker-bake.hcl --set '*.platform=linux/ppc64le' --push linux-ppc64le
```

Capture d’écran 2023-05-14 à 13 00 06

- Then, logged in as DockerHub `jenkins` account with "write" permission, I've updated the existing manifest by appending the ppc64le from the `dduportal` image:

```
$ docker buildx imagetools inspect dduportal/jenkins:2.403-jdk17
Name:      docker.io/dduportal/jenkins:2.403-jdk17
MediaType: application/vnd.oci.image.index.v1+json
Digest:    sha256:d5102e8a0ec3f2ddcf76441e9161bb4ace07a747fe8388f441910760f687b5a7

Manifests:
  Name:        docker.io/dduportal/jenkins:2.403-jdk17@sha256:b464824f1fefc51ef1afcb3f9db23d045c468afff667d206b15b8eb13d96f292
  MediaType:   application/vnd.oci.image.manifest.v1+json
  Platform:    linux/ppc64le

  Name:        docker.io/dduportal/jenkins:2.403-jdk17@sha256:d444c81a33ad0ca1c6e719868a13d8e816539c605c6f2abe5a0bfa8f486c9f6d
  MediaType:   application/vnd.oci.image.manifest.v1+json
  Platform:    unknown/unknown
  Annotations:
    vnd.docker.reference.digest: sha256:b464824f1fefc51ef1afcb3f9db23d045c468afff667d206b15b8eb13d96f292
    vnd.docker.reference.type:   attestation-manifest
$ docker buildx imagetools inspect jenkins/jenkins:2.403-jdk17
Name:      docker.io/jenkins/jenkins:2.403-jdk17
MediaType: application/vnd.oci.image.index.v1+json
Digest:    sha256:9a6fbd7bbaeec025d6cb080a8c0165adf475084c62a1ad8e04f910256cd577e8

Manifests:
  Name:        docker.io/jenkins/jenkins:2.403-jdk17@sha256:155becf87d9dc093a2ac70cec503df1a470587c70fafb51459fe981451f8b845
  MediaType:   application/vnd.oci.image.manifest.v1+json
  Platform:    linux/amd64

  Name:        docker.io/jenkins/jenkins:2.403-jdk17@sha256:87884045aba234543be145e951a810ae96448b6486bc5be9b812c2f77400008b
  MediaType:   application/vnd.oci.image.manifest.v1+json
  Platform:    linux/arm64

  Name:        docker.io/jenkins/jenkins:2.403-jdk17@sha256:50c06e487d4e5d3afd33035a7af27c1a54268b723c6e56d220e8e01c3b9ba342
  MediaType:   application/vnd.oci.image.manifest.v1+json
  Platform:    unknown/unknown
  Annotations:
    vnd.docker.reference.digest: sha256:155becf87d9dc093a2ac70cec503df1a470587c70fafb51459fe981451f8b845
    vnd.docker.reference.type:   attestation-manifest

  Name:        docker.io/jenkins/jenkins:2.403-jdk17@sha256:8295a71789f806de957fb9d5f93489e93ec996296008151019b4766f49bd3ea8
  MediaType:   application/vnd.oci.image.manifest.v1+json
  Platform:    unknown/unknown
  Annotations:
    vnd.docker.reference.digest: sha256:87884045aba234543be145e951a810ae96448b6486bc5be9b812c2f77400008b
    vnd.docker.reference.type:   attestation-manifest

$ docker buildx imagetools create --append --dry-run jenkins/jenkins:2.403-jdk17 dduportal/jenkins:2.403-jdk17
{
  "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
  "schemaVersion": 2,
  "manifests": [
    {
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "digest": "sha256:155becf87d9dc093a2ac70cec503df1a470587c70fafb51459fe981451f8b845",
      "size": 2577,
      "platform": {
        "architecture": "amd64",
        "os": "linux"
      }
    },
    {
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "digest": "sha256:87884045aba234543be145e951a810ae96448b6486bc5be9b812c2f77400008b",
      "size": 2577,
      "platform": {
        "architecture": "arm64",
        "os": "linux"
      }
    },
    {
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "digest": "sha256:50c06e487d4e5d3afd33035a7af27c1a54268b723c6e56d220e8e01c3b9ba342",
      "size": 566,
      "annotations": {
        "vnd.docker.reference.digest": "sha256:155becf87d9dc093a2ac70cec503df1a470587c70fafb51459fe981451f8b845",
        "vnd.docker.reference.type": "attestation-manifest"
      },
      "platform": {
        "architecture": "unknown",
        "os": "unknown"
      }
    },
    {
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "digest": "sha256:8295a71789f806de957fb9d5f93489e93ec996296008151019b4766f49bd3ea8",
      "size": 566,
      "annotations": {
        "vnd.docker.reference.digest": "sha256:87884045aba234543be145e951a810ae96448b6486bc5be9b812c2f77400008b",
        "vnd.docker.reference.type": "attestation-manifest"
      },
      "platform": {
        "architecture": "unknown",
        "os": "unknown"
      }
    },
    {
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "digest": "sha256:b464824f1fefc51ef1afcb3f9db23d045c468afff667d206b15b8eb13d96f292",
      "size": 2577,
      "platform": {
        "architecture": "ppc64le",
        "os": "linux"
      }
    },
    {
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "digest": "sha256:d444c81a33ad0ca1c6e719868a13d8e816539c605c6f2abe5a0bfa8f486c9f6d",
      "size": 566,
      "annotations": {
        "vnd.docker.reference.digest": "sha256:b464824f1fefc51ef1afcb3f9db23d045c468afff667d206b15b8eb13d96f292",
        "vnd.docker.reference.type": "attestation-manifest"
      },
      "platform": {
        "architecture": "unknown",
        "os": "unknown"
      }
    }
  ]
}

docker buildx imagetools create --append jenkins/jenkins:2.403-jdk17 dduportal/jenkins:2.403-jdk17 --tag=jenkins/jenkins:2.403-jdk17
[+] Building 4.2s (1/1) FINISHED
 => [internal] pushing docker.io/jenkins/jenkins:2.403-jdk17
```

Capture d’écran 2023-05-14 à 13 02 44

**Comment by dduportal** [comment]
Update: manual publication has the expected effect on our "custom" publication script so we should be able to proceed with this method:

```
# Get list of tags
$ docker buildx bake -f docker-bake.hcl --set '*.platform=linux/ppc64le' linux-ppc64le --print | jq -r '.target[].tags[]'
docker.io/dduportal/jenkins:2.403
docker.io/dduportal/jenkins:2.403-jdk11
docker.io/dduportal/jenkins:2.403-jdk17             # Already done previously
docker.io/dduportal/jenkins:2.403-rhel-ubi9-jdk17
# For each tag:

$ docker buildx imagetools inspect --raw jenkins/jenkins:2.403-jdk11
$ docker buildx imagetools create --append --dry-run jenkins/jenkins:2.403-jdk11 dduportal/jenkins:2.403-jdk11 --tag=jenkins/jenkins:2.403-jdk11
# Compare output with previous one: should only append ppc64le from dduportal/jenkins
$ docker buildx imagetools create --append jenkins/jenkins:2.403-jdk11 dduportal/jenkins:2.403-jdk11 --tag=jenkins/jenkins:2.403-jdk11
# Should succeed and check the result in DockerHub

$ docker buildx imagetools inspect --raw jenkins/jenkins:2.403
$ docker buildx imagetools create --append --dry-run jenkins/jenkins:2.403 dduportal/jenkins:2.403 --tag=jenkins/jenkins:2.403
# Compare output with previous one: should only append ppc64le from dduportal/jenkins
$ docker buildx imagetools create --append jenkins/jenkins:2.403 dduportal/jenkins:2.403 --tag=jenkins/jenkins:2.403
# Should succeed and check the result in DockerHub

$ docker buildx imagetools inspect --raw jenkins/jenkins:2.403-rhel-ubi9-jdk17
$ docker buildx imagetools create --append --dry-run jenkins/jenkins:2.403-rhel-ubi9-jdk17 dduportal/jenkins:2.403-rhel-ubi9-jdk17 --tag=jenkins/jenkins:2.403-rhel-ubi9-jdk17
# Compare output with previous one: should only append ppc64le from dduportal/jenkins
$ docker buildx imagetools create --append jenkins/jenkins:2.403-rhel-ubi9-jdk17 dduportal/jenkins:2.403-rhel-ubi9-jdk17 --tag=jenkins/jenkins:2.403-rhel-ubi9-jdk17
# Should succeed and check the result in DockerHub

# Check the effect on our "custom publication script":

$ ./.ci/publish.sh -n
Docker repository in Use:
* JENKINS_REPO: jenkins/jenkins
Dry run, will not publish images

Tag is already published: 2.403

# ... Other are seen as not publish for now
```

**Comment by dduportal** [comment]
TODO list

Update these image manifest to append with ppc64le:

- [x] 2.403 (previous weekly)
- [ ] 2.404 (latest weekly)
- [ ] 2.387.2 (previous LTS)
- [ ] 2.387.3 (latest LTS)

[ ] Then check that `./.ci/publish.sh -n` from this PR returns the 4 following lines:

```
Tag is already published: 2.403
Tag is already published: 2.404
Tag is already published: 2.387.2
Tag is already published: 2.387.3
```

=> then we can merge

**Comment by MarkEWaite** [comment]
I think that we should update the changelogs of the [weekly](https://www.jenkins.io/changelog/) and [LTS releases](https://www.jenkins.io/changelog-stable/) that will now have ppc64le container images so that they note the addition of ppc64le as a supported container architecture.  I'm happy to submit the necessary pull request if others are OK with that idea.

**Comment by ksalerno99** [comment]
> I think that we should update the changelogs of the [weekly releases](https://www.jenkins.io/changelog/) and [LTS releases](https://www.jenkins.io/changelog-stable/) that will now have ppc64le container images so that they note the addition of ppc64le as a supported container architecture. I'm happy to submit the necessary pull request if others are OK with that idea.

Mark, thank you. Also the file HACKING.adoc also needs to be updated to list ppc64le as a supported arch as well.

**Comment by dduportal** [comment]
> I think that we should update the changelogs of the [weekly releases](https://www.jenkins.io/changelog/) and [LTS releases](https://www.jenkins.io/changelog-stable/) that will now have ppc64le container images so that they note the addition of ppc64le as a supported container architecture. I'm happy to submit the necessary pull request if others are OK with that idea.

For this particular case, I would treat the manual operations to previous weekly and LTS as a "shadow" launch and would wait for tuesday (16 May) weekly to add it to the changelog.
Rationale is that adding manually the PPC64le does not change the existing images. With this PR being merged soon, before the 2.405 weekly, the changelog would be easier to dig for users if they can relate to a code change which timeline make sense.
=> WDYT @MarkEWaite ?

**Comment by MarkEWaite** [comment]
> For this particular case, I would treat the manual operations to previous weekly and LTS as a "shadow" launch and would wait for tuesday (16 May) weekly to add it to the changelog.

That sounds good to me.

> Rationale is that adding manually the PPC64le does not change the existing images.

Maybe I misunderstood something you said earlier.  I thought that the 2.404 container images for all architectures would get new SHA values (as a side effect) when the ppc64le container images are published for them.  I'd love to not have those released container images changed if that is possible, but I accept that they are changed if that is not possible.

**Comment by dduportal** [comment]
> > For this particular case, I would treat the manual operations to previous weekly and LTS as a "shadow" launch and would wait for tuesday (16 May) weekly to add it to the changelog.
> 
> That sounds good to me.
> 
> > Rationale is that adding manually the PPC64le does not change the existing images.
> 
> Maybe I misunderstood something you said earlier. I thought that the 2.404 container images for all architectures would get new SHA values (as a side effect) when the ppc64le container images are published for them. I'd love to not have those released container images changed if that is possible, but I accept that they are changed if that is not possible.

For the 2.403, it's what happenned alas, due to a misunderstanding I have about how Docker Buildx works.

But then, the second attempt in adding ppc64le to 2.403 showed that we can add new images with the `docker buildx imagetools` command without changing the existing ones.
=> that is what I plan to do to the latest weekly 2.404, as well as the 2 past LTS, e.g. "only adding" ppc64le architecure along to the existing architectures, without changing their SHAs.

**Review by dduportal — APPROVED** [review]
```
./.ci/publish.sh -n                 
Docker repository in Use:
* JENKINS_REPO: jenkins/jenkins
Dry run, will not publish images
Tag is already published: 2.403
Tag is already published: 2.404
Tag is already published: 2.387.2
Tag is already published: 2.387.3
```

LGTM ! Thanks @ksalerno99

**Pull request merged**